pkg/validators/validator_001_default_channel.go

```diff
@@ -0,0 +1,91 @@
+package validators
+
+import (
+	"fmt"
+
+	"github.com/mt-sre/addon-metadata-operator/api/v1alpha1"
+	"github.com/mt-sre/addon-metadata-operator/pkg/utils"
+)
+
+// ValidateDefaultChannel validates whether the 'defaultChannel' provided under an addon.yaml also exists under 'channels' field
+func ValidateDefaultChannel(metabundle utils.MetaBundle) (bool, string, error) {
+	defaultChannel := metabundle.AddonMeta.DefaultChannel
+	channels := metabundle.AddonMeta.Channels
+
+	if success, failureMsg := isPartOfEnum(defaultChannel); !success {
+		return false, failureMsg, nil
+	}
+	// to be deprecated - only used for legacy builds
+	if success, failureMsg := isListedInChannels(channels, defaultChannel); !success {
+		return false, failureMsg, nil
+	}
+
+	// if success, failureMsg, err := matchesDefaultChannelAnnotations(metabundle.Bundles, defaultChannel); !success {
+	// 	return false, failureMsg, err
+	// }
+	return true, "", nil
+}
+
+func isPartOfEnum(defaultChannel string) (bool, string) {
+	enum := map[string]struct{}{
+		"alpha":  {},
+		"beta":   {},
+		"stable": {},
+		"edge":   {},
+		"rc":     {},
+	}
+	if _, ok := enum[defaultChannel]; !ok {
+		return false, fmt.Sprintf("The defaultChannel '%v' is not part of the accepted values: alpha, beta, stable, edge or rc.", defaultChannel)
+	}
+	return true, ""
+}
+
+// TODO - deprecate this when we remove legacy builds
+func isListedInChannels(channels *[]v1alpha1.Channel, defaultChannel string) (bool, string) {
+	// as the Channels field is deprecated, it can be omitted
+	if channels == nil {
+		return true, ""
+	}
+	var channelNames []string
+	for _, channel := range *channels {
+		if channel.Name == defaultChannel {
+			return true, ""
+		}
+		channelNames = append(channelNames, channel.Name)
+	}
+	return false, fmt.Sprintf("The defaultChannel '%v' is not part of the listed channelNames: %v.", defaultChannel, channelNames)
+}
+
+// TODO - (sblaisdo) enable when extraction format is bundles instead of packageManifest
+// if the annotation is not present, make sure the defaultChannel is alpha
+// func matchesDefaultChannelAnnotations(bundles []registry.Bundle, defaultChannel string) (bool, string, error) {
+// 	for _, bundle := range bundles {
+// 		allChannels := strings.Split(bundle.Annotations.Channels, ",")
```

Comment:
```suggestion
//               if bundle.Annotations == nil {
//                   continue
//               }
// 		allChannels := strings.Split(bundle.Annotations.Channels, ",")
```

https://github.com/mt-sre/addon-metadata-operator/pull/34#discussion_r768602631

> The if Bundle.Annotations because without a nil-check this might trigger a nil-dereference panic if bundle.Annotations doesn't exist. As Annotations is a ptr and behind the scenes, bundle.Annotations.Channels would be auto-dereferenced to access Annotations.Channels
 
:)

Reply:
good point. I fixed it, but we expect the extractor to fail if there are no annotations.yaml files, or if there is any problem reading it (not the case right now). I reflected that in the code.

Reply:
Gotcha! Thanks :)